### PR TITLE
Clean up User-defined type related parser flag

### PR DESCRIPTION
**Pull request by adithyaselv**
* Remove `AllowParseAsTypeLiteral, AllowTypeLiteral` parser flag.
* AsType, IsType and ParseJSON overloads are part of Builtin Functions
* Add constant for Type Keyword.
* Add type arg functions to Built-in
* Refactor test

### CODE REVIEW

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.DefinitionsCheckResult.#ctor

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.DefinitionsCheckResult.#ctor

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.DefinitionsCheckResult.#ctor

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.DefinitionsCheckResult.#ctor

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.DefinitionsCheckResult.#ctor

File: src/libraries/Microsoft.PowerFx.Core/Public/DefinitionsCheckResult.cs [hunk 1]
```diff
@@ -49,9 +51,10 @@ public class DefinitionsCheckResult : IOperationStatus
         // All errors accumulated. 
         private readonly List<ExpressionError> _errors = new List<ExpressionError>();
 
-        public DefinitionsCheckResult()
+        public DefinitionsCheckResult(Features features = null)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/DefinitionsCheckResult.cs:54]: >DefinitionsCheckResult [](http://example.com/codeflow?start=15&length=22)

nit - if you were trying to do this in a non-breaking way (see warnings in PR) 

you could add a new ctor, and have existing one chain to it:

DefinitionsCheckResult() : this(Features.PowerFxV1)

DefinitionsCheckResult(Features features) { }

File: src/libraries/Microsoft.PowerFx.Core/Public/DefinitionsCheckResult.cs [hunk 2]
```diff
@@ -49,9 +51,10 @@ public class DefinitionsCheckResult : IOperationStatus
...
         {
             _localSymbolTable = new SymbolTable { DebugName = "LocalUserDefinitions" };
+            _features = features ?? Features.None;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/DefinitionsCheckResult.cs:57]: > Features.None; [](http://example.com/codeflow?start=35&length=15)

Can we default to v1? And let PA fall back to v0?

Comment (adithyaselv) [reply src/libraries/Microsoft.PowerFx.Core/Public/DefinitionsCheckResult.cs]: using V1 for Definitions but still keeping None for Format and other uses, as TexlParser defaults to None

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 3]
```diff
@@ -90,16 +87,16 @@ private TexlParser(IReadOnlyList<Token> tokens, Flags flags, Features features =
         /// </summary>
         /// <param name="script">Script to be parsed.</param>
         /// <param name="parserOptions">Options for parsing an expression.</param>
+        /// <param name="features">Power Fx feature flags.</param>
         /// <returns><see cref="ParseUserDefinitionResult"/>.</returns>
-        public static ParseUserDefinitionResult ParseUserDefinitionScript(string script, ParserOptions parserOptions)
+        public static ParseUserDefinitionResult ParseUserDefinitionScript(string script, ParserOptions parserOptions, Features features = null)
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:92]: >ParserOptions parserOptions, Features feature [](http://example.com/codeflow?start=89&length=45)

We have a PFxV1 in both ParserOptions and in Features... Not something that needs to be fixed in this PR, but we should have a follow-up issue to clean this up.

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 4]
```diff
@@ -2048,11 +2045,13 @@ internal static string GetTokString(TokKind kind)
         /// </summary>
         /// <param name="text">Expression text to format.</param>
         /// <param name="flags">Optional flags to customize the behavior of underlying lexer and parser. By default, expression chaining is enabled.</param>
+        /// <param name="features">Power Fx features.</param>
         /// <returns>Formatted expression text.</returns>
-        public static string Format(string text, Flags flags = Flags.EnableExpressionChaining)
+        public static string Format(string text, Flags flags = Flags.EnableExpressionChaining, Features features = null)
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:2050]: >Features features = null [](http://example.com/codeflow?start=95&length=24)

Not a blocker, but does it need to be optional?

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt [hunk 5]
```diff
@@ -25,21 +23,6 @@ true
 >> AsType(ParseJSON("""1984-01-01"""), Date)
 Date(1984,1,1)
 
->> AsType(ParseJSON("""1900-12-31T23:59:59.999Z"""), DateTime)
-DateTime(1900,12,31,15,59,59,999)
-
->> AsType(ParseJSON("""1900-12-31T23:59:59.999"""), DateTime)
-DateTime(1900,12,31,23,59,59,999)
-
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:33]: This doesn't need to be specific to the US Pacific time zone

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO_TimeZone_Seattle.txt [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#SETUP: TimeZoneInfo("Pacific Standard Time")
+
+>> AsType(ParseJSON("""1900-12-31T23:59:59.999Z"""), DateTime)
+DateTime(1900,12,31,15,59,59,999)
+
+>> AsType(ParseJSON("""1900-12-31T23:59:59.999"""), DateTime)
+DateTime(1900,12,31,23,59,59,999)
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO_TimeZone_Seattle.txt:7]: This doesn't need to be specific to the US Pacific time zone

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO_TimeZone_Seattle.txt [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+>> AsType(ParseJSON("""1900-12-31T23:59:59.999+00:00"""), DateTime)
+DateTime(1900,12,31,15,59,59,999)
+
+>> AsType(ParseJSON("""1900-12-31T23:59:59.999-08:00"""), DateTime)
+DateTime(1900,12,31,23,59,59,999)
+
+>> AsType(ParseJSON("""1900-12-31"""), DateTime)
+DateTime(1900,12,31,0,0,0,0)
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO_TimeZone_Seattle.txt:16]: This doesn't need to be specific to the US Pacific time zone

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/TypedParseJSON_TimeZone_Seattle.txt [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#SETUP: TimeZoneInfo("Pacific Standard Time")
+
+>> ParseJSON("""2008-01-01T12:12:12.100Z""", DateTime)
+DateTime(2008,1,1,4,12,12,100)
+
+>> ParseJSON("""2008-01-01T12:12:12.100""", DateTime)
+DateTime(2008,1,1,12,12,12,100)
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/TypedParseJSON_TimeZone_Seattle.txt:7]: This doesn't need to be specific to the US Pacific time zone

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/TypedParseJSON_TimeZone_Seattle.txt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+>> ParseJSON("""2008-01-01T12:12:12.100-08:00""", DateTime)
+DateTime(2008,1,1,12,12,12,100)
+
+>> ParseJSON("""1900-12-31""", DateTime)
+DateTime(1900,12,31,0,0,0,0)
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/TypedParseJSON_TimeZone_Seattle.txt:13]: This doesn't need to be specific to the US Pacific time zone

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs (hunk 4 above)
Comment (adithyaselv) [reply src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:2050]: It does not need to be but added for flexibility. I see that the API for Format and FormatUserDefinitions can we made a bit consistent and improved , created an issue to handle these cleanups - https://github.com/microsoft/Power-Fx/issues/2706

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**